content/post/postgres/extension-preloading.md

```diff
@@ -0,0 +1,170 @@
+---
+title: Extension Preloading
+slug: extension-preloading
+date: 2024-07-15T21:03:01Z
+lastMod: 2024-07-15T21:03:01Z
+description: When should a Postgres extension be pre-loaded and when should it not?
+tags: [Postgres, Extensions, Preload]
+type: post
+---
+
+Recently I've been trying to figure out when a Postgres extension module
+should be preloaded. By "extension" I mean, more generally, shared libraries
+provided or used by extensions to Postgres, whether a `CREATE EXTENSION`
+extension written in C or [pgrx] or a [`LOAD`]able module. By "preloaded" I
+mean under what conditions should it be added to one of the [Shared Library
+Preloading] variables:
+
+*   `local_preload_libraries`
+*   `session_preload_libraries`
+*   `shared_preload_libraries`
+
+The answer, it turns out, comes very much down to the extension type.
+
+Normal Extensions
+-----------------
+
+If your extension includes no loadable modules, congratulations! You don't
+have to worry about this question at all.
+
+If your extension's shared library provides functionality only via functions
+called from SQL, you also don't need to worry about this preloading. Custom
+types, operators, and functions generally follow this pattern. The DDL that
+creates the SQL object, such as [`CREATE FUNCTION`], uses the `AS 'obj_file',
+'link_symbol'` syntax to tell that PostgreSQL what library to load when it's
+needed.
+
+Hook Extensions
+---------------
+
+If your extension's shared library makes calls to PostgreSQL without
+PostgreSQL first calling it, then the library must be loaded before it's used.
+This is typically the case for libraries that modifies the server's behavior
+through "hooks" rather than providing a set of functions.
+
+### Session Preloading
+
+If your extension is intended for debugging or performance-measurement, it
+likely doesn't need to be preloaded for every connection. In this case, a DBA
+might allow specific users to load it by either:
+
+*   Adding it to the [`session_preload_libraries`] variable for the user via
+    [`ALTER ROLE`], so it loads for every connection for that user:
+
+    ```sql
+    ALTER ROLE role_name
+      SET session_preload_libraries TO '$libdir/mylib';
+    ```
+
+*   Granting the user role the ability to set `session_preload_libraries`
```

Comment:
This feels wrong.  If the user role_name logs in, their loaded libraries will be those specified.  If they don't log in directly  (options are not inherited) the presence of this doesn't give them the ability to add it.  I haven't tried to find evidence either way though.  Do you have a reference or example of this functioning?